Source/Caches/HACacheKeyStates.swift

```diff
@@ -27,22 +27,23 @@ internal struct HACacheKeyStates: HACacheKey {
     static func processUpdates(
         info: HACacheTransformInfo<HACompressedStatesUpdates, HACachedStates?>,
         shouldResetEntities: Bool
-    )
-        -> HACachedStates {
-        var states: HACachedStates
-        if shouldResetEntities {
-            states = .init(entities: [])
-        } else {
-            states = info.current ?? .init(entities: [])
+    ) -> HACachedStates {
+        var updatedEntities: [HAEntity] = []
+
+        if !shouldResetEntities, let currentEntities = info.current?.all {
+            updatedEntities = Array(currentEntities)
         }
+
         if let additions = info.incoming.add {
             for (entityId, updates) in additions {
-                if var currentState = states[entityId] {
+                if var currentState = updatedEntities.first(where: { $0.entityId == entityId }),
+                   let index = updatedEntities.firstIndex(of: currentState) {
```

Comment:
This is now pretty bad at performance - at least two O(N) added on top of another O(N) enumeration. It's also going to require another step to segment by dictionary again at the end, which is a lot of overhead for all of these changes.

I think it may be better to make the 'all' access lazy, or give HACachedStates an initializer that takes a dictionary, or something along those lines.

Reply:
Shame on me, I did the opposite of the PR title :P 

I just updated it, I went for dealing with a dictionary for the whole process to avoid wasting by look inside an array and endup creating the cache in the end with an initializer that receives a dictionary as you said